### ISSUE TITLE
New command line options to change the size of an existing shared cache

### ISSUE THREAD

**Issue by hangshao0** [eclipse/openj9#7494]
Currently there is no efficient way to determine the appropriate size of the shared cache for an application. The process involves trail and error, which means users have to startup/shut down their application several times and check the cache statistics to figure out the correct shared cache size. This is undesirable. 

We can introduce new command line utility options `-Xshareclasses:correctCacheSize and -Xshareclasses:correctCacheSize=<size>` that copy the content of an existing shared cache to a new cache file with the correct size and delete the old cache file. Thus, user can start up their application only once with a shared cache that is large enough, then use `correctCacheSize` to shrink the shared cache to the appropriate size.

The layout of the shared cache is:
 https://github.com/eclipse/openj9/blob/6c7b5015c3c7b702bbe26377bc75406beeeb912b/runtime/shared_common/CompositeCache.cpp#L48toL56

Only the size of free space will be changed by `correctCacheSize`. Any references to addresses after the free space (metadata area and class debug data area) using offset to the cache head need to be fixed.

**Comment by hangshao0** [comment]
> Any references to addresses after the free space (metadata area and class debug data area) using offset to the cache head need to be fixed.

@mpirvu Does this break JIT/AOT data that are stored to the share cache ?

**Comment by pshipton** [comment]
I think we should pick a different name for the option. Like minimizeCacheSize or shrinkCacheSize. Calling it `correct` implies that it wasn't correct before, which depends on the way it's being used.

**Comment by hangshao0** [comment]
Or probably `adjustCacheSize`, as user might want to increase the cache size to store more data.

**Comment by DanHeidinga** [comment]
I'd vote for two options:
* `minimizeCacheSize` to shrink the cache to the smallest possible size for the current data
* `resizeCache` to grow or shrink the cache to a specific size.  The smallest allowed resize would be equal tot he `minimizeCacheSize` size.

**Comment by mpirvu** [comment]
>> Any references to addresses after the free space (metadata area and class debug data area) using offset to the cache head need to be fixed.

> Does this break JIT/AOT data that are stored to the shared cache ?

As long as ROMClasses in SCC keep the same offsets we should be fine

**Comment by hangshao0** [comment]
I've updated the description to a design. FYI: @pshipton @DanHeidinga

**Comment by pshipton** [comment]
- J9ShrOffset is used to refer to both metadata and classes/methods. It doesn't help to use end of metadata instead of cache start, unless this is only done for metadata pointers, which should be doable.

- Even if the debug area is at the start of the cache, there is still the free space to deal with. If the free space is removed, some SRPs will need updates.

- The read/write area can be removed when the cache is 100% full (minimizeCacheSize).

- Depending on the final solution (if necessary), resizing the cache should modify the unique ID to invalidate higher layers which have pointers into the cache.

**Comment by hangshao0** [comment]
> unless this is only done for metadata pointers, which should be doable.

Yes. Only for metadata pointers.

> Even if the debug area is at the start of the cache, there is still the free space to deal with. If the free space is removed, some SRPs will need updates.

> The read/write area can be removed when the cache is 100% full (minimizeCacheSize).

Rom methods are using SRPs pointing to LVT, which is after the free space in the debug data area. I don't see rom methods pointing to LNT. So if the debug area is at the start of the cache, the relative position of LVT and rom methods won't be affected by the size of free space in the debug data area. 

If we allow read/write area to be removed, the cache layout could be like this:

 | Header | RW area | LNT | <-  resizable freespace -> | LVT | ROM segment | <-  resizable freespace -> | metadata |

> resizing the cache should modify the unique ID to invalidate higher layers which have pointers into the cache.

Yes. fileSize and createtime are part of the unique ID. They will be changed once a cache is resized.

**Comment by pshipton** [comment]
Methods have to point into the LNT. 
https://github.com/eclipse/openj9/blob/171820baf418b10a317bdc90606758dec76b4760/runtime/util/mthutil.c#L299-#L315

**Comment by hangshao0** [comment]
OK, then change the layout does not seem to help much. We still need to fix the SRPs for each rom method if its debug data are stored out-of-line. Also all the SRPs used in struct J9VariableInfoValues needs to be fixed. Not sure if we want to do that. It could be a lot of pointers to fix.  

Probably forcing the debug data to be stored inline is an easier solution, though the debug data won't be disclaimed in this case.

**Comment by pshipton** [comment]
We want to keep the debug data on disk instead of memory. The new layout works to avoid most of the fixups. The pointers to the line number tables will still need to be fixed, but that's not hard to do.

**Comment by hangshao0** [comment]
Just talked to Peter, we could go with the layout in https://github.com/eclipse/openj9/issues/7494#issuecomment-582109322, so that we don't need to change SRPs used in struct J9VariableInfoValues. But we still need to fix the pointer to the LNT for each rom method.

**Comment by hangshao0** [comment]
I am looking at the implementation. I found some issue with removing the readWrite area:

1. If SH_CacheMap::storeSharedData() is called with J9SHRDATA_USE_READWRITE, we will allocate in the readWrite area. The corresponding meta data will point to the allocated data in the readWrite area. So if the readWrite area is removable, I guess we need to stop supporting J9SHRDATA_USE_READWRITE. I don't find any caller using this flag now in the our code though. 
https://github.com/eclipse/openj9/blob/master/runtime/shared_common/CacheMap.cpp#L4053

2. What happens if the user runs `minimizeCacheSize` first (readWrite area removed), and later increase the cache size ? We return an increased cache with no readWrite area ?

**Comment by pshipton** [comment]
Agreed, considering J9SHRDATA_USE_READWRITE has never been used, it can be removed.

> What happens if the user runs minimizeCacheSize first (readWrite area removed), and later increase the cache size ? We return an increased cache with no readWrite area ?

The readWrite area will need to be re-created, or never removed in the first place. The readWrite area is needed for utf8 sharing between ROM classes. If it's removed the table of most used utf8's is gone and cannot be recovered. Although every individual JVM will also do utf8 sharing, but will only build a table from newly loaded classes, not classes that are already loaded in the cache. I suppose removing the readWrite area could be a separate option. I see a readWrite size of 363936 bytes for the default cache size. If the cache doesn't touch anything in this area it won't be loaded from disk, perhaps it's enough to not load anything from the readWrite area if the cache is full.

**Comment by pshipton** [comment]
Actually, I forgot a case. Every JVM should still be able to shared utf8's with the shared cache, even if the cache is full. If this is correct, we shouldn't remove the readWrite area.

**Comment by hangshao0** [comment]
OK, I will proceed with not removing the readWrite area now. 

Also I noticed that the debug region should be placed before rather than after the readWrite area. In this way the readWrite area is next to the ROM segment, and their relative position won't be affected by resizing.  Updated in the description.

**Comment by dsouzai** [comment]
Debug Counter data is stored into the SCC using
```
   dataDescriptor.type    = J9SHR_DATA_TYPE_JITHINT;
   dataDescriptor.flags   = J9SHRDATA_NOT_INDEXED;
```
AOT Class chains are stored into the SCC using
```
   dataDescriptor.type    = J9SHR_DATA_TYPE_AOTCLASSCHAIN;
   dataDescriptor.flags   = J9SHRDATA_SINGLE_STORE_FOR_KEY_TYPE;
```
The relocation record then holds the offset into the SCC. We use the offset to get hold of the Debug Counter data or Class Chain. If the offset from the start of the SCC can change, then this will break existing AOT.

**Comment by hangshao0** [comment]
The romclass offsets will remain the same as they are offset to cache head. See:
https://github.com/eclipse/openj9/issues/7494#issuecomment-543327139

But if you are using offsets to cache head to locate data stored by AOT/JIT in SCC. Then it will be broken as they are stored after the free space. Instead, you can use offset to cache end or SRP.

**Comment by DanHeidinga** [comment]
It sounds like we need to add some test cases that validate the structures in the cache are still valid after a resize operation.  Not sure what those tests would look like but something that ensures we have all the appropriate meta data, romclasses, aot code, class chains, etc and that they are well formed before and after the operation would make it easier to evolve this code.

**Comment by dsouzai** [comment]
> But if you are using offsets to cache head to locate data stored by AOT/JIT in SCC. Then it will be broken as they are stored after the free space. Instead, you can use offset to cache end or SRP.

The problem is that we use the same APIs to convert between pointers and offsets for both ROMClasses and AOT/JIT data. I can't just switch the APIs to calculate from the end of the cache because then the offsets of the ROMClasses will be incorrect.

I'll have to create a second set of APIs for the AOT/JIT data, and then go through the code and make sure we use the right APIs in the right places. This will likely take some time.

**Comment by dsouzai** [comment]
JIT support has been added in https://github.com/eclipse/openj9/pull/10369
